backend/server/src/nix/mod.rs

```diff
@@ -0,0 +1,50 @@
+use anyhow::Result;
+use std::process::Command;
+use tracing::{debug, info};
+use std::collections::HashMap;
+
+/// Given a drv, traverse all direct drv dependencies
+pub fn traverse_drvs (drv_path: &str) -> Result<HashMap<String, Vec<String>>> {
+    debug!("traversing {}", drv_path);
+
+    let mut drv_to_references: HashMap::<String, Vec<String>> = HashMap::new();
+    inner_traverse_drvs(&mut drv_to_references, drv_path)?;
+
+    debug!("drvs {:?}", drv_to_references);
+    Ok(drv_to_references)
+}
+
+/// To avoid lifetime issues, we do a recursive descent instead of a loop
+/// instead of appending to an iterator :(
+fn inner_traverse_drvs(drv_map: &mut HashMap<String, Vec<String>>, drv_path: &str) -> Result<()> {
+    let references = drv_references(drv_path)?;
+    drv_map.insert(drv_path.to_string(), references.clone());
+
+    for drv in references.into_iter() {
+        if drv_map.contains_key(&drv) {
```

Comment:
This logic should query the global drv cache, and then we would only be traversing the delta of new drvs

Reply:
For time being, this map is meant as a memoization mechanism so I'm not descending down shared drv dependencies. In the future, the workflow will be something like:
- determine which attrs have changed
  - for each each attr, determine all new drvs (this PR tackles this)
  - enqueue each new drv as a build request
- later, have a build queue/scheduler attempt to build the related drvs